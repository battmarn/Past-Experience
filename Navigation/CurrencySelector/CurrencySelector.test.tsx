import React from 'react';

import { act, userEvent } from 'app/bit-migration/test-utilities/react-renderer';
import { createUseCurrentLocationMock } from 'app/hooks/useCurrentLocationMock';
import { clearDataLayer, snapshotDataLayer } from 'app/lib/analytics/analytics.mock';
import { createUseTrackingLocationMock } from 'app/providers/GTM/TrackingLocationProvider.mock';
import { renderWithStore } from 'test/helpers/renderWithStore';

import { withNavigationContext } from '../Context/Context';

import { CurrencySelector as CurrencySelectorComponent } from '.';

const CurrencySelector = withNavigationContext(CurrencySelectorComponent);

describe('CurrencySelector', () => {
  beforeEach(() => {
    clearDataLayer();
    createUseCurrentLocationMock();
    createUseTrackingLocationMock('Header - Menu');
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should render as expected', () => {
    const { container } = renderWithStore(<CurrencySelector />);
    expect(container).toMatchSnapshot();
  });

  it('should behave as expected when elements are hovered over', async () => {
    const { getByText, getByTestId } = renderWithStore(<CurrencySelector />);

    expect(getByTestId('country-select')).toHaveAttribute('aria-hidden', 'true');

    await userEvent.hover(getByTestId('header-selected-site'), { delay: null });

    expect(getByTestId('country-select')).toHaveAttribute('aria-hidden', 'false');

    await userEvent.click(getByText('SEK | kr'), { delay: null });

    snapshotDataLayer(2);

    expect(getByTestId('header-selected-site')).toHaveTextContent('SEK | kr');
  });

  it('should fire events on open and close - mouse', async () => {
    const { getByTestId } = renderWithStore(<CurrencySelector />);

    expect(getByTestId('country-select')).toHaveAttribute('aria-hidden', 'true');

    // hover
    await userEvent.pointer({ target: getByTestId('header-selected-site') }, { delay: null });

    expect(getByTestId('country-select')).toHaveAttribute('aria-hidden', 'false');

    // then click the existing hovered item
    await userEvent.pointer({ target: getByTestId('header-selected-site'), keys: '[MouseLeft]' }, { delay: null });

    expect(getByTestId('country-select')).toHaveAttribute('aria-hidden', 'true');

    // then click it again
    await userEvent.pointer({ target: getByTestId('header-selected-site'), keys: '[MouseLeft]' }, { delay: null });

    expect(getByTestId('country-select')).toHaveAttribute('aria-hidden', 'false');

    snapshotDataLayer(3);
  });

  it('should fire events on open and close - touch', async () => {
    const { getByTestId } = renderWithStore(<CurrencySelector />);

    expect(getByTestId('country-select')).toHaveAttribute('aria-hidden', 'true');

    // tap on selector to open
    await userEvent.pointer({ target: getByTestId('header-selected-site'), keys: '[TouchA]' }, { delay: null });

    expect(getByTestId('country-select')).toHaveAttribute('aria-hidden', 'false');

    // tap on selector again to close
    await userEvent.pointer({ target: getByTestId('header-selected-site'), keys: '[TouchA]' }, { delay: null });

    expect(getByTestId('country-select')).toHaveAttribute('aria-hidden', 'true');

    snapshotDataLayer(2);
  });

  it('when mouse leaves allow 300ms before closing the menu', async () => {
    jest.useFakeTimers();

    const { getByTestId } = renderWithStore(<CurrencySelector />);
    const selectedSiteHeader = getByTestId('header-selected-site');
    const menu = getByTestId('country-select');

    expect(menu).toHaveAttribute('aria-hidden', 'true');

    await userEvent.hover(selectedSiteHeader, { delay: null });

    expect(menu).toHaveAttribute('aria-hidden', 'false');

    await userEvent.unhover(selectedSiteHeader, { delay: null });

    expect(menu).toHaveAttribute('aria-hidden', 'false');

    act(() => {
      jest.advanceTimersByTime(300);
    });

    expect(menu).toHaveAttribute('aria-hidden', 'true');

    snapshotDataLayer(2);
  });
});
