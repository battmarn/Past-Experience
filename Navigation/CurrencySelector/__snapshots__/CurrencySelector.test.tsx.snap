// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySelector should behave as expected when elements are hovered over 1`] = `
Array [
  Object {
    "event": "site_interaction",
    "site_interaction_info": Object {
      "featureType": undefined,
      "location": "Header - Menu - Currency Selector",
      "name": "Opened",
    },
  },
  Object {
    "event": "site_interaction",
    "site_interaction_info": Object {
      "featureType": "click",
      "location": "Header - Menu - Currency Selector",
      "name": "Selected new site - se",
    },
  },
]
`;

exports[`CurrencySelector should fire events on open and close - mouse 1`] = `
Array [
  Object {
    "event": "site_interaction",
    "site_interaction_info": Object {
      "featureType": undefined,
      "location": "Header - Menu - Currency Selector",
      "name": "Opened",
    },
  },
  Object {
    "event": "site_interaction",
    "site_interaction_info": Object {
      "featureType": undefined,
      "location": "Header - Menu - Currency Selector",
      "name": "Closed",
    },
  },
  Object {
    "event": "site_interaction",
    "site_interaction_info": Object {
      "featureType": undefined,
      "location": "Header - Menu - Currency Selector",
      "name": "Opened",
    },
  },
]
`;

exports[`CurrencySelector should fire events on open and close - touch 1`] = `
Array [
  Object {
    "event": "site_interaction",
    "site_interaction_info": Object {
      "featureType": undefined,
      "location": "Header - Menu - Currency Selector",
      "name": "Opened",
    },
  },
  Object {
    "event": "site_interaction",
    "site_interaction_info": Object {
      "featureType": undefined,
      "location": "Header - Menu - Currency Selector",
      "name": "Closed",
    },
  },
]
`;

exports[`CurrencySelector should render as expected 1`] = `
<div>
  <div
    class="border"
    role="menu"
    tabindex="0"
  >
    <div
      aria-hidden="true"
      class="desktopSelect hidden"
      data-test-id="country-select"
      role="menu"
      tabindex="0"
    >
      <p
        class="menuShippingText"
      >
        site-menu-header
      </p>
      <a
        class="menuOption"
        data-test-id="desktop-currency-menu-option"
        href="/de"
        target="_top"
      >
        <img
          alt="uk flag"
          class="image"
          src="/assets/flags/de.svg"
        />
        EUR | €
      </a>
      <a
        class="menuOption"
        data-test-id="desktop-currency-menu-option"
        href="/se"
        target="_top"
      >
        <img
          alt="uk flag"
          class="image"
          src="/assets/flags/se.svg"
        />
        SEK | kr
      </a>
      <a
        class="menuOption active"
        data-test-id="desktop-currency-menu-option"
        href="/uk"
        target="_top"
      >
        <img
          alt="uk flag"
          class="image"
          src="/assets/flags/uk.svg"
        />
        GBP | £
      </a>
    </div>
    <div
      class="flagsContainer"
    >
      <button
        class="buttonSelect"
        data-test-id="desktop-currency-button-select"
      >
        <img
          alt="uk flag"
          class="image"
          src="/assets/flags/uk.svg"
        />
        <span
          data-test-id="header-selected-site"
        >
          GBP | £
        </span>
        <img
          alt="arrow"
          class="arrows"
          src="test-file-stub"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`CurrencySelector when mouse leaves allow 300ms before closing the menu 1`] = `
Array [
  Object {
    "event": "site_interaction",
    "site_interaction_info": Object {
      "featureType": undefined,
      "location": "Header - Menu - Currency Selector",
      "name": "Opened",
    },
  },
  Object {
    "event": "site_interaction",
    "site_interaction_info": Object {
      "featureType": undefined,
      "location": "Header - Menu - Currency Selector",
      "name": "Closed",
    },
  },
]
`;
